rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+                self.uncompressed.mark_written(res.wrote);
+                self.compressed.mark_read(res.consumed);
+                chunk_state.compressed_remaining -= res.consumed as u64;
+                if self.compressed.len() == 0 {
+                    self.compressed.reset();
+                }
+                match self.uncompressed.span(amount) {
+                    Span(b) => Ok(Span(b)),
+                    Remainder(_) => Ok(Remainder(std::cmp::min(
+                        clamp_to_usize(chunk_state.compressed_remaining),
+                        res.need,
+                    ))),
+                }
+            }
+        }
+    }
+
+    // Consume `amount` bytes of the uncompressed input buffer if enough is available. On failure,
+    // return the extra amount required as an error value.
+    fn consume(&mut self, amount: usize) -> McapResult<SpanOrRemainder> {
+        match self.load(amount)? {
+            Span(span) => {
+                self.uncompressed.mark_read(amount);
+                Ok(Span(span))
+            }
+            Remainder(remainder) => Ok(Remainder(remainder)),
+        }
+    }
+
+    // Return an InputBuf that requests `want` bytes from the input file. If reading
+    // from a chunk, reads into the compressed buffer, otherwise reads into the uncompressed buffer.
+    fn request(&mut self, want: usize) -> McapResult<Option<ReadAction>> {
+        return match &self.from {
+            File => Ok(Some(ReadAction::Fill(InputBuf {
+                buf: self.uncompressed.tail_with_size(want),
+                last_write: &mut self.last_write,
+            }))),
+            Chunk(state) => {
+                if state.decompressor.is_none() {
+                    Ok(Some(ReadAction::Fill(InputBuf {
+                        buf: self.uncompressed.tail_with_size(want),
+                        last_write: &mut self.last_write,
+                    })))
+                } else {
+                    Ok(Some(ReadAction::Fill(InputBuf {
+                        buf: self.compressed.tail_with_size(want),
+                        last_write: &mut self.last_write,
+                    })))
+                }
+            }
+        };
+    }
+}
+
+impl Default for LinearReader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Encapsulates the action the user should take next when reading an MCAP file.
+///
+/// ```no_run
+/// use mcap::sans_io::read::ReadAction;
+/// use mcap::McapResult;
+/// use mcap::read::parse_record;
+/// fn use_action(file: &mut Box<dyn std::io::Read>, action: ReadAction) -> McapResult<()> {
+///   match action {
+///     ReadAction::Fill(mut into) => {
+///       let n = file.read(into.buf)?;
+///       into.set_filled(n);
+///     },
+///     ReadAction::GetRecord{ opcode, data } => {
+///       let record = mcap::parse_record(opcode, data)?;
+///       // do something with the record...
+///     }
+///   }
+///   Ok(())
+/// }
+/// ```
+pub enum ReadAction<'a> {
```

Comment:
trying to make these verbs feels a bit weird – maybe something like `ReadResult`? and the case names would be something like `Record` & `Incomplete` or `NeedsMore`?

Reply:
I'm open to that - let me try renaming and see how it feels.